### ISSUE TITLE
Baseline information for contextual layout

### ISSUE THREAD

**Issue by litherum** [w3c/svgwg#739]
Forked from https://github.com/w3c/csswg-drafts/issues/4116. That issue describes motivation to be able to annotate images with information which can participate in layout in surrounding context (like CSS). At TPAC this year in Tokyo, I brought up this topic in the SVG CG, and received encouragement to make a proposal which would add this kind of information in SVG.

The most straightforward way of doing this modeling this layout information is as an additional baseline. Indeed, this is how UIKit [models](https://developer.apple.com/documentation/uikit/uiimage/3375727-baselineoffsetfrombottom) this information.

There are three general approaches to baseline information: [OpenType](https://docs.microsoft.com/en-us/typography/opentype/spec/baselinetags), [TrueType](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6bsln.html), and [CSS](https://drafts.csswg.org/css-inline-3/#dominant-baseline-property). These describe how there can be a collection of different baselines: Roman, Mathematical, Hanging, etc. Also, OpenType includes facilities for vertical baselines, which are helpful for aligning things horizontally.

So, gathering inspiration from `viewBox`, I was thinking of a microsyntax approach. Perhaps something like two new attributes on the `<svg>` element: `horizontal-baselines` and vertical-baselines`, which would look something similar to:

```
<svg horizontal-baselines=“‘hang’ 1014 ‘romn’ 14 ‘ideo’ 589” vertical-baselines=“‘icfb’ -142 ‘idtp’ 820 ‘romn’ -3”> ... </svg>
```

I got those 4 character codes from the [OpenType spec](https://docs.microsoft.com/en-us/typography/opentype/spec/baselinetags)  but we may want to prefer the CSS spec instead, in which case it would be something like:

```
<svg horizontal-baselines=“hanging 1014 alphabetic 14 ideographic 589” vertical-baselines=“middle -142 mathematical 820 alphabetic -3”> ... </svg>
```

**Comment by litherum** [comment]
(This issue is supposed to be in the SVG CG but I can’t figure out how to open issues there)

**Comment by heycam** [comment]
Using the human readable CSS baseline names seems preferable to me.

**Comment by faceless2** [comment]
At risk of reiterating the suggestion I made in https://github.com/w3c/csswg-drafts/issues/4116, it seems unnecessarily complicated where the baseline you are aligning to is designed to match a text element within the SVG. That text element will already have font metrics.

Your proposal is more complete and covers every situation, including those where there's no text element to reference. For the use-cases you've mentioned elsewhere - svg's representing logos or symbols - the syntax is as good as any. I too prefer the CSS baseline names. And I presume the values are in viewBox units?

But for anything more complex, it's going to be an extremely arduous job to calculate those numbers.  Take something like, for example, http://web-platform-tests.live/svg/import/text-align-07-t-manual.svg. Calculating the alignment points manually in advance essentially requires the user to pre-calculate the linebox vertical alignment by hand, an algorithm which still surprises me with the results - and that's after I've implemented it. I'm sure an SVG editor could calculate them for inclusion, but you're then shifting the burden of development from just SVG renderers to both SVG renderers and SVG editors.

At the very least I would suggest at least an alternate syntax which allows you to retrieve those metrics from an embedded <text> or <tspan> by reference. I proposed one syntax in https://github.com/w3c/csswg-drafts/issues/4116 (`<svg baselineElement="textid">`) - the precise syntax could be changed to work with your proposal quite easily, eg `<svg horizontal-baselines="#textid">`.

Even for simple icons containing no text content, this would allow - for example - a stub <text> element with the right metrics but no text content to be included, just to copy the metrics from.

**Comment by litherum** [comment]
That’s a good point! You’re absolutely right.

**Comment by litherum** [comment]
What about if we add a valueless attribute to one of the descendants of the <svg>, along with some resolution rules about how to handle if it’s on two distinct elements? That way, we don’t need to resolve an additional URL.

We also need to consider what happens if the representative element is inside the <use> shadow tree, and is potentially duplicated throughout the SVG.

**Comment by faceless2** [comment]
Thank you, I hope it has some merit. I'm not quite sure I follow the idea of the valueless attribute - do you mean the item used for the external baseline identifies itself by some sort of marker attribute? Like this sort of thing?
```xml
<svg>
 <path d="..."/> <!-- complex path defining the shape -->
 <text i-am-the-horizontal-baseline="whatever" font="48px Times-Roman" x="0" y="22"/>
</svg>
```
I suppose there's not much between the two approaches. When you're linking into an HTML document with a fragment, and there are two items with the same fragment, the first takes precedence. The same approach would make sense here.

As for within the shadow tree, you mean this sort of case?
```xml
<svg>
 <defs>
  <text id="foo" i-am-the-horizontal-baseline="whatever" font="48px Times-Roman"/>
 </defs>
 <path d="..."/> <!-- complex path defining the shape -->
 <use id="bar" href="#foo" x="20" y="30"/>
 <use id="baz" href="#foo" x="20" y="50"/>
</svg>
```
In this case, if the element is announcing itself, it's a bit ambiguous - which copy do you want? The first seems to make the most sense.

Alternatively if the element is being referenced (say with  `<svg horizontal-baseline="foo">` then I think there's no ambiguity: the item with id "foo" is inside a `<defs>` which has `display: none`, so it has no position and therefore no baseline. Which, I guess, means there is an advantage to the use of a valueless attribute over referencing one by id.

**Comment by faceless2** [comment]
To give this some context - one of my goals would be to make this sort of structure possible (photo taken from a children's book)

![pathtext](https://user-images.githubusercontent.com/989243/65518628-83633880-dedc-11e9-8a5a-998488e3a790.jpg)

Which, I would imagine, you would be able to create with something like this:

```xml
<p>
 It sounds
 <svg style="width:content-fit; height: content-fit; alignment-baseline: alphabetic">
  <text i-am-the-horizontal-baseline="">
   <textPath path="..."><tspan class="large">utterly dreary</tspan> to me</textPath>
  </text>
 </svg>
 <br/>
 When I get home I...
</p>
```
(Note I'm also anticipating an as-yet-unspecced "content-fit" value for SVG width and height which would shrink the SVG to the bounds of its visible content)

**Comment by heycam** [comment]
It's not really clear what the baseline of text that consists solely of a `<textPath>` would be.  In this specific example, it looks like you want to align to the y coordinate of the starting point of the path.  Since that value must be known -- it'll be in the path data -- you can provide it explicitly in the `horizontal-baseline=""` attribute on the root `<svg>`.

(You are right though that there's no way to size the inline `<svg>` to its contents automatically.)

**Comment by faceless2** [comment]
@heycam - yes, fair point. I was making the assumption that the baseline of the textpath is the baseline at its start. One option would be to allow "start" or "end" to be specified - perhaps as a value to the previously-valueless attribute, or perhaps by using a structure like this:
```xml
<text>
 <textPath path="...">
  <tspan class="large">utterly dreary</tspan> to me
  <tspan i-am-the-horizontal-baseline="">&zwsp;</tspan>
 </textPath>
</text>
```

If the text is itself subject to a transform or within a nested viewBox etc, then calculating that "y" value is not necessarily trivial - a statement that applies to any child element, not just text.

In addition, the "y" value specified for <text> is the Y value of the dominant baseline - alphabetic, in my example above. If I understand @litherum's suggestion, the idea is to specify all the other baselines too - hanging, mathematical and so on.

In short: if there's a transform or viewbox, or you want to specify a baseline other than the dominant one, working out the Y value(s) has to be done by hand and is potentially not trivial.
